plugin/impl/hypervisor.go

```diff
@@ -0,0 +1,172 @@
+// SPDX-FileCopyrightText: 2020 SAP SE or an SAP affiliate company
+// SPDX-License-Identifier: Apache-2.0
+
+package impl
+
+import (
+	"fmt"
+	"reflect"
+	"strings"
+
+	v1 "github.com/cobaltcore-dev/openstack-hypervisor-operator/api/v1"
+	"github.com/sapcc/ucfgwrap"
+	"k8s.io/apimachinery/pkg/types"
+	"k8s.io/client-go/util/retry"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+
+	"github.com/sapcc/maintenance-controller/plugin"
+)
+
+// CheckHypervisor is a check plugin, which checks properties of the hypervisor status CRO of the node.
+type CheckHypervisor struct {
+	Fields *map[string]any
+}
+
+func (i *CheckHypervisor) OnTransition(params plugin.Parameters) error {
+	return nil
+}
+
+// New creates a new CheckHypervisor instance with the given config.
+func (i *CheckHypervisor) New(config *ucfgwrap.Config) (plugin.Checker, error) {
+	conf := struct {
+		Fields map[string]any `config:",inline"`
+	}{}
+	if err := config.Unpack(&conf); err != nil {
+		return nil, err
+	}
+
+	mappedFields, err := mapToStructFields(conf.Fields, reflect.TypeFor[v1.HypervisorStatus]())
+	if err != nil {
+		return nil, err
+	}
+	return &CheckHypervisor{&mappedFields}, nil
+}
+
+func (i *CheckHypervisor) ID() string {
+	return "checkHypervisor"
+}
+
+// Check checks whether the hypervisor is evicted.
+func (i *CheckHypervisor) Check(params plugin.Parameters) (plugin.CheckResult, error) {
+	var hypervisor v1.Hypervisor
+	if err := params.Client.Get(params.Ctx, types.NamespacedName{Name: params.Node.Name}, &hypervisor); err != nil {
+		return plugin.Failed(nil), err
+	}
+
+	for key, value := range *i.Fields {
+		s := reflect.ValueOf(&hypervisor.Status).Elem()
+		f := s.FieldByName(key)
+		if !f.IsValid() {
+			return plugin.Failed(nil), fmt.Errorf("field %s not found in Hypervisor status", key)
+		}
+
+		switch v := value.(type) {
+		case bool:
+			if f.Bool() != v {
+				return plugin.Failed(nil), nil
+			}
+		case string:
+			if f.String() != v {
+				return plugin.Failed(nil), nil
+			}
+		case []string:
+			fieldSlice, ok := f.Interface().([]string)
+			if !ok {
+				return plugin.Failed(nil), fmt.Errorf("field %s is not of type []string", key)
+			}
+			if !reflect.DeepEqual(fieldSlice, v) {
+				return plugin.Failed(nil), nil
+			}
```

Comment:
This code will panic if the type of `f` is not equal to the type of `v`.

Reply:
couldn't reproduce, I've added a test for type mismatch.